File: balancer/service/login/src/core/InserviceService.cc
#include "InserviceService.h"

InserviceList::InserviceList(unsigned short service_id, const std::vector<ServiceConfig::IPInfo>& il)
	: depend_service_id(service_id), count(0), inservice_list(il)
{
	for(auto& ip_info: inservice_list)
	{
		inservice_map.insert(std::make_pair(ip_info.proc_id, ip_info));
	}
}

bool InserviceService::load_ip_info(unsigned short service_id, ServiceConfig& sc)
{
	auto p = sc.get_service(service_id);
	if(p == nullptr)
	{
		return _is_ok = false;	//缺少本服务的配置
	}

	//遍历本服务的依赖服务ipinfo
	for(auto it = p->depend_map.begin(); it != p->depend_map.end(); it++)
	{
		auto p_depend_service = sc.get_service(it->second.depend_service_id);
		if(p_depend_service == nullptr || p_depend_service->inservice_list.empty())
		{
			return _is_ok = false; //没有依赖服务的ipinfo
		}

		_map.insert(std::make_pair(it->second.depend_service_id, 
			InserviceList(it->second.depend_service_id, p_depend_service->inservice_list)));
	}

	return _is_ok = _kv.load_kv(p->kv_map);
}

ServiceConfig::IPInfo* InserviceService::get_ip_info(unsigned short depend_service_id)
{
	auto it = _map.find(depend_service_id);
	if(it != _map.end())
	{
		InserviceList& il = it->second;
		unsigned int size = il.inservice_list.size();
		if(size > 0)
		{
			unsigned int index = (++il.count) % size;
			return &il.inservice_list[index];
		}
	}

	return nullptr;
}

ServiceConfig::IPInfo* InserviceService::get_ip_info(unsigned short depend_service_id, unsigned int proc_id)
{
	auto it = _map.find(depend_service_id);
	if(it != _map.end())
	{
		InserviceList& il = it->second;
		auto it_ip_info = il.inservice_map.find(proc_id);
		if(it_ip_info != il.inservice_map.end())
		{
			return &it_ip_info->second;
		}
	}

	return nullptr;
}

InserviceList* InserviceService::get_inservice_list(unsigned short depend_service_id)
{
	auto it = _map.find(depend_service_id);
	if(it != _map.end())
	{
		return &it->second;
	}

	return nullptr;
}

File: balancer/service/proxy/src/core/inservice_service.py
# -*- coding: utf-8 -*-

from src.core import kv

class InserviceList:

    def __init__(self, service_id, il):
        self.depend_service_id = service_id
        self.count = 0
        self.inservice_list = il
        self.inservice_map = {}
        for ip_info in self.inservice_list:
            self.inservice_map[ip_info.proc_id] = ip_info


class InserviceService:

    def __init__(self):
        self.is_ok = False
        self.map = {}
        self.kv = kv.Kv()

    def load_ip_info(self, service_id, sc):
        service = sc.get_service(service_id)
        if not service:
            self.is_ok = False
            return self.is_ok

        for depend in service.depend_map:
            depend_service = sc.get_service(depend.depend_service_id)
            if depend_service is None or len(depend_service.inservice_list) == 0:
                self.is_ok = False
                return self.is_ok

            il = InserviceList(depend.depend_service_id, 
                               depend_service.inservice_list)
            self.map[depend.depend_service_id] = il

        self.is_ok = self.kv.load_kv(service.kv_map)
        return self.is_ok

    def get_ip_info(self, depend_service_id, proc_id=0):
        il = self.map.get(depend_service_id)
        if il is not None:
            if proc_id == 0:
                size = len(il.inservice_list)
                if size > 0:
                    il.count += 1
                    index = il.count % size
                    return il.inservice_list[index]
            else:
                return il.inservice_map.get(proc_id)

        return None

    def get_inservice_list(self, depend_service_id):
        return self.map.get(depend_service_id)





